# Mini Project: Build a Machine Learning Model

## Predict Total Fare on the NYC Taxi Dataset

Welcome to the NYC Taxi Fare Prediction project! In this Colab, we will continue using the NYC Taxi Dataset to predict the fare amount for taxi rides using a subset of available features. We will go through three main stages: building a baseline model, creating a full model, and performing hyperparameter tuning to enhance our predictions.

The previous EDA work on the NYC Taxi Dataset can be found in the [16.3.3. Student MLE MiniProject EDA](https://github.com/dmkimura/mec2-projects/blob/main/16.3.3_Student_MLE_MiniProject_EDA.ipynb) Jupyter notebook.

Now that you've completed exploratory data analysis on this dataset you should have a good understanding of the feature space.

## Project Objectives

The primary objectives of this project are as follows:

Baseline Model: We will start by building a simple baseline model to establish a benchmark for our predictions. This model will serve as a starting point to compare the performance of our subsequent models.

Full Model: Next, we will develop a more comprehensive model that leverages machine learning techniques to improve prediction accuracy. We will use Scikit-Learn's model pipeline to build a framework that enables rapid experimentation.

Hyperparameter Tuning: Lastly, we will optimize our full model by fine-tuning its hyperparameters. By systematically adjusting the parameters that control model behavior, we aim to achieve the best possible performance for our prediction task.

In [55]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import mean_absolute_error

from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler, OneHotEncoder, FunctionTransformer
from sklearn.compose import ColumnTransformer
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor

Load the NYC taxi dataset into a Pandas DataFrame and do a few basic checks to ensure the data is loaded properly. Note, there are several months of data that can be used. For simplicity, use the Yellow Taxi 2022-01 parquet file [here](https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2022-01.parquet). Here are your tasks:

  1. Load the `yellow_tripdata_2022-01.parquet` file into Pandas.
  2. Print the first 5 rows of data.
  3. Drop any rows of data that contain NULL values.
  4. Create a new feature, 'trip_duration' that captures the duration of the trip in minutes.
  5. Create a varible named 'target_variable' to store the name of the thing we're trying to predict, 'total_amount'.
  6. Create a list called 'feature_cols' containing the feature names that we'll be using to predict our target variable. The list should contain 'VendorID', 'trip_distance', 'payment_type', 'PULocationID', 'DOLocationID', and 'trip_duration'.

In [56]:
# Load the dataset into a pandas DataFrame (from https://www.nyc.gov/site/tlc/about/tlc-trip-record-data.page)
df = pd.read_parquet(
    'https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2022-01.parquet')

In [57]:
# Display the first few rows of the dataset
df.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         1  2022-01-01 00:35:40   2022-01-01 00:53:29              2.0   
1         1  2022-01-01 00:33:43   2022-01-01 00:42:07              1.0   
2         2  2022-01-01 00:53:21   2022-01-01 01:02:19              1.0   
3         2  2022-01-01 00:25:21   2022-01-01 00:35:23              1.0   
4         2  2022-01-01 00:36:48   2022-01-01 01:14:20              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0           3.80         1.0                  N           142           236   
1           2.10         1.0                  N           236            42   
2           0.97         1.0                  N           166           166   
3           1.09         1.0                  N           114            68   
4           4.30         1.0                  N            68           163   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             1         14.5    3.0      0.5        3.65           0.0   
1             1          8.0    0.5      0.5        4.00           0.0   
2             1          7.5    0.5      0.5        1.76           0.0   
3             2          8.0    0.5      0.5        0.00           0.0   
4             1         23.5    0.5      0.5        3.00           0.0   

   improvement_surcharge  total_amount  congestion_surcharge  airport_fee  
0                    0.3         21.95                   2.5          0.0  
1                    0.3         13.30                   0.0          0.0  
2                    0.3         10.56                   0.0          0.0  
3                    0.3         11.80                   2.5          0.0  
4                    0.3         30.30                   2.5          0.0

### Data Cleaning

Remove missing values.

In [58]:
# Drop rows with missing values.
print(f'Current missing values:\n{df.isna().sum()}')

df.dropna(inplace=True)

print(f'\nAfter removing missing values:\n{df.isna().sum()}')
#df.head()

Current missing values:
VendorID                     0
tpep_pickup_datetime         0
tpep_dropoff_datetime        0
passenger_count          71503
trip_distance                0
RatecodeID               71503
store_and_fwd_flag       71503
PULocationID                 0
DOLocationID                 0
payment_type                 0
fare_amount                  0
extra                        0
mta_tax                      0
tip_amount                   0
tolls_amount                 0
improvement_surcharge        0
total_amount                 0
congestion_surcharge     71503
airport_fee              71503
dtype: int64

After removing missing values:
VendorID                 0
tpep_pickup_datetime     0
tpep_dropoff_datetime    0
passenger_count          0
trip_distance            0
RatecodeID               0
store_and_fwd_flag       0
PULocationID             0
DOLocationID             0
payment_type             0
fare_amount              0
extra                    0
mta_tax           

### Feature engineering.
* create new feature called 'trip_duration'
* remove datetime features

**Create new feature**

The new feature will calculate the duration of each taxi ride in minutes based on the tpep_dropoff_datetime and tpep_pickup_datetime features.

In [59]:
# Create new feature, 'trip_duration'.

# from previous EDA, columns with datetime values are in UTC time
# will convert datetime values from UTC to New York time zone
df['tpep_pickup_datetime'] = df[
    'tpep_pickup_datetime'].dt.tz_localize('America/New_York', ambiguous='NaT')

df['tpep_dropoff_datetime'] = df[
    'tpep_dropoff_datetime'].dt.tz_localize('America/New_York', ambiguous='NaT')

# create variable with trip duration based on drop off and pickup times
trip_duration = df['tpep_dropoff_datetime'] - df['tpep_pickup_datetime']

# create new df column for trip duration expressed in minutes
df['trip_duration'] = trip_duration.dt.total_seconds() / 60

print(df.head(3))

   VendorID      tpep_pickup_datetime     tpep_dropoff_datetime  \
0         1 2022-01-01 00:35:40-05:00 2022-01-01 00:53:29-05:00   
1         1 2022-01-01 00:33:43-05:00 2022-01-01 00:42:07-05:00   
2         2 2022-01-01 00:53:21-05:00 2022-01-01 01:02:19-05:00   

   passenger_count  trip_distance  RatecodeID store_and_fwd_flag  \
0              2.0           3.80         1.0                  N   
1              1.0           2.10         1.0                  N   
2              1.0           0.97         1.0                  N   

   PULocationID  DOLocationID  payment_type  fare_amount  extra  mta_tax  \
0           142           236             1         14.5    3.0      0.5   
1           236            42             1          8.0    0.5      0.5   
2           166           166             1          7.5    0.5      0.5   

   tip_amount  tolls_amount  improvement_surcharge  total_amount  \
0        3.65           0.0                    0.3         21.95   
1        4.00    

**Remove datetime features**

Datetime datatypes are not useable datatypes for machine learning models. Since the two datetime features in the dataset were used to create a new feature, these features can be excluded from the X data used for training and testing models.

Let's create a list containing all of the features that will be used as the X dataset for model training and testing.

In [60]:
# Create a list called feature_col to store column names
target = 'total_amount'
remove_col = [target, 'tpep_dropoff_datetime', 'tpep_pickup_datetime']
feature_col = [col for col in df.columns if col not in remove_col]
feature_col

['VendorID',
 'passenger_count',
 'trip_distance',
 'RatecodeID',
 'store_and_fwd_flag',
 'PULocationID',
 'DOLocationID',
 'payment_type',
 'fare_amount',
 'extra',
 'mta_tax',
 'tip_amount',
 'tolls_amount',
 'improvement_surcharge',
 'congestion_surcharge',
 'airport_fee',
 'trip_duration']

### Train, Test, Split

Splitting a dataset into training and test sets is a crucial step in machine learning model development. It allows us to evaluate the performance and generalization ability of our models accurately. The training set is used to train the model, while the test set serves as an independent sample for evaluating its performance.

1. **Model Training**: The training set is used to fit the model, allowing it to learn the underlying patterns and relationships between the features and the target variable. By exposing the model to a diverse range of examples in the training set, it can capture the underlying structure of the data.

2. **Model Evaluation**: The test set, which is independent of the training set, is crucial for evaluating how well the trained model generalizes to unseen data. It provides an unbiased assessment of the model's performance on new instances. By measuring the model's accuracy, precision, recall, or other evaluation metrics on the test set, we can estimate how well the model will perform on unseen data.

3. **Preventing Overfitting**: Overfitting occurs when a model learns the training data's noise and idiosyncrasies instead of the underlying patterns. By evaluating the model on the test set, we can identify if the model is overfitting. If the model performs significantly worse on the test set compared to the training set, it indicates overfitting. In such cases, we might need to adjust the model, feature selection, or regularization techniques to improve generalization.

4. **Hyperparameter Tuning**: Splitting the dataset allows us to perform hyperparameter tuning on the model. Hyperparameters are configuration settings that control the learning process, such as learning rate, regularization strength, or the number of hidden layers in a neural network. By using a validation set (often created from a portion of the training set), we can iteratively adjust the hyperparameters and select the best combination that maximizes the model's performance on the validation set. The final evaluation on the test set provides an unbiased estimate of the model's performance.

By splitting the dataset into training and test sets, we can ensure that our models are both well-trained and accurately evaluated. This separation helps us understand how the model will perform on new, unseen data, which is critical for assessing its effectiveness and making informed decisions about its deployment.

Here is your task:

  1. Use Scikit-Learn's [train_test_split](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html) to split the data into training and test sets. Don't forget to set the random state.

In [61]:
# see current shape of the dataframe
df.shape

(2392428, 20)

In [62]:
# Split dataset into 70% training and 30% test sets

X = df[feature_col]
y = df[target]

X_test, X_train, y_test, y_train = train_test_split(X, y, test_size=0.3, random_state=1337)

### Baseline Model

The importance of a baseline model, even if it uses a simple strategy like always predicting the mean, cannot be understated. Here's why a baseline model is valuable:

1. **Performance Comparison**: A baseline model serves as a reference point for evaluating the performance of more sophisticated models. By establishing a simple yet reasonable baseline, we can determine whether our advanced models offer any significant improvement over this basic approach. It helps us set realistic expectations and gauge the effectiveness of our efforts.

2. **Model Complexity**: A baseline model provides insight into the complexity required to solve the prediction task. If a simple strategy like predicting the median performs reasonably well, it suggests that the problem might not necessitate complex modeling techniques. Conversely, if the baseline model performs poorly, it indicates the presence of more intricate patterns that need to be captured by more sophisticated models.

3. **Minimum Performance Requirement**: A baseline model can establish a minimum performance requirement for a predictive task. If we cannot outperform the baseline, it suggests that our models have failed to capture even the most fundamental relationships within the data. In such cases, we may need to revisit our data preprocessing steps, feature engineering techniques, or consider other external factors affecting the task.

4. **Identifying Data Issues**: A baseline model can help identify potential issues within the dataset. If the baseline model performs poorly, it may indicate problems like missing values, outliers, or data inconsistencies. These issues can be further investigated and resolved to improve the overall model performance.

While a baseline model like always predicting the median may not offer the highest prediction accuracy, its importance lies in its role as a starting point for model development and evaluation. It provides a solid foundation for comparing and assessing the performance of more complex models, ensuring that any improvements made are meaningful and significant.

Here is your task:

  1. Create a model that always predicts the mean total fare of the training dataset. Use Scikit-Learn's [mean_absolute_error](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.mean_absolute_error.html) to evaluate this model. Is it any good?

In [9]:
# Create a baseline model that predicts the mean of the training set
from sklearn.dummy import DummyRegressor

dum_reg = DummyRegressor(strategy='mean')

dum_reg.fit(X_train, y_train)

y_pred = dum_reg.predict(X_test)

# show that the model is predicting the mean of the
# target values in the training set
print(f'The mean value of the y_train set:\n{y_train.mean()}')
print(f'\nThe target values predicted by the baseline model:\n{y_pred}')

The mean value of the y_train set:
18.834789398226917

The target values predicted by the baseline model:
[18.8347894 18.8347894 18.8347894 ... 18.8347894 18.8347894 18.8347894]


In [10]:
# calculate the mean absolute error for the baseline model
base_mae = mean_absolute_error(y_test, y_pred)
base_mae

9.332187766854899

### Linear Regression Model

With a baseline metric in place, we can try to build a machine learning model. Obviously, if the model can't beat the baseline then there are some major issues to be resolved.

It's always a good idea to start with a simple machine learning model, like linear regression, and build upon it if necessary.

Here are your tasks:

  1. Use Scikit-Learn's [ColumnTransformer](https://scikit-learn.org/stable/modules/generated/sklearn.compose.ColumnTransformer.html) to preprocess the categorical and continuous features independently. Apply the [StandardScaler](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html) to the continuous columns and [OneHotEncoder](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OneHotEncoder.html) to the categorical columns.

  One-hot encoding is a popular technique used to represent categorical variables numerically in machine learning models. It transforms categorical features into a binary vector representation, where each category is represented by a binary column. Here's an explanation of one-hot encoding:

  When working with categorical variables, such as colors (e.g., red, blue, green) or vehicle types (e.g., car, truck, motorcycle), machine learning algorithms often require numerical inputs. However, directly assigning numerical values to categories can introduce unintended relationships or orderings between them. For example, assigning the values 0, 1, and 2 to the categories red, blue, and green may imply a sequential relationship, which is not desired.

  One-hot encoding solves this problem by creating new binary columns, equal to the number of unique categories in the original feature. Each binary column represents a specific category and takes a value of 1 if the data point belongs to that category, and 0 otherwise. This encoding ensures that no implicit ordering or relationship exists between the categories.

  2. Integrate the preprocessor in the previous step with Scikit-Learn's [LinearRegression](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html) model using a [Pipeline](https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html).

  3. Train the pipeline on the training data.

  4. Evaluate the model using mean absolute error as a metric on the test data. Does the model beat the baseline?


**Use ColumnTransformer to Preprocess Data**

In [11]:
# create list of feature names with categorial values
cat_col = list(X.select_dtypes(include=['object']).columns)
cat_col

['store_and_fwd_flag']

In [12]:
# create list of feature names with continuous values
cont_col = list(X.select_dtypes(exclude=['object']).columns)
cont_col

['VendorID',
 'passenger_count',
 'trip_distance',
 'RatecodeID',
 'PULocationID',
 'DOLocationID',
 'payment_type',
 'fare_amount',
 'extra',
 'mta_tax',
 'tip_amount',
 'tolls_amount',
 'improvement_surcharge',
 'congestion_surcharge',
 'airport_fee',
 'trip_duration']

In [13]:
# Use Scikit-Learn's ColumnTransformer to preprocess the categorical and
# continuous features independently.
ct = ColumnTransformer(
    [("categorical", OneHotEncoder(), cat_col),
     ("continuous", StandardScaler(), cont_col)
     ]
    )

**Create pipeline to perform preprocessing and model training**

In [14]:
# Create a pipeline object containing the column transformations and regression
# model.
linr_pipe = Pipeline([('preprocessor', ct),
                    ('regressor', LinearRegression())]
            )

**Use pipeline to train the data**

In [15]:
# Fit the pipeline on the training data.
linr_pipe.fit(X_train, y_train)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('categorical',
                                                  OneHotEncoder(),
                                                  ['store_and_fwd_flag']),
                                                 ('continuous',
                                                  StandardScaler(),
                                                  ['VendorID',
                                                   'passenger_count',
                                                   'trip_distance',
                                                   'RatecodeID', 'PULocationID',
                                                   'DOLocationID',
                                                   'payment_type',
                                                   'fare_amount', 'extra',
                                                   'mta_tax', 'tip_amount',
                                                   'tolls_amount',
                                                   'improvement_surcharge',
                                                   'congestion_surcharge',
                                                   'airport_fee',
                                                   'trip_duration'])])),
                ('regressor', LinearRegression())])

**Evaluate the model using mean absolute error**

In [16]:
# Make predictions on the test data.
y_linr_pred = linr_pipe.predict(X_test)

# calculate mean absolute error on the predictions
linr_mae = mean_absolute_error(y_test, y_linr_pred)

In [17]:
# display results
print(f"Baseline model's mae:\t\t{base_mae:.5f}")
print(f"Linear Regression's mae:\t{linr_mae:.5f}")

Baseline model's mae:		9.33219
Linear Regression's mae:	0.13891


The above linear regression model clearly performs better than the baseline line model. The closer the mae is to zero, the better the model performs.

### Random Forest Regression

Random Forest Regression and Linear Regression are two commonly used regression algorithms, each with its own advantages and suitability for different scenarios. Random Forest Regression offers several advantages over Linear Regression, including:

1. **Non-linearity**: Random Forest Regressor is capable of capturing non-linear relationships between features and the target variable. In contrast, Linear Regression assumes a linear relationship between the features and the target. When faced with non-linear relationships or complex feature interactions, Random Forest Regressor can provide more accurate predictions.

2. **Robustness to Outliers**: Random Forest Regressor is generally more robust to outliers compared to Linear Regression. Outliers can disproportionately impact the coefficients and predictions of Linear Regression models. However, as an ensemble of decision trees, Random Forest Regressor can mitigate the effect of outliers by averaging predictions from multiple trees.

3. **Feature Importance**: Random Forest Regressor provides a measure of feature importance, which helps identify the most influential features for making predictions. This information is useful for feature selection, understanding the underlying relationships in the data, and gaining insights into the problem domain. Unlike Linear Regression, which provides coefficient values indicating the direction and magnitude of relationships, Random Forest Regressor explicitly highlights feature importance.

4. **Handling of Categorical Variables**: Random Forest Regressor can effectively handle categorical variables without requiring pre-processing steps like one-hot encoding. It can directly incorporate categorical variables into the model, making it more convenient when working with mixed data types. In contrast, Linear Regression often requires categorical variables to be encoded or transformed before use.

5. **Handling of High-Dimensional Data**: Random Forest Regressor can handle datasets with a large number of features (high dimensionality) by automatically selecting subsets of features during the construction of individual decision trees. This reduces the risk of overfitting, which is a concern with Linear Regression when dealing with high-dimensional data.

6. **Resistance to Multicollinearity**: Random Forest Regressor is less affected by multicollinearity, which occurs when predictor variables are highly correlated. In Linear Regression, highly correlated features can lead to unstable coefficient estimates, making it challenging to interpret the individual effects of each feature. Random Forest Regressor, as an ensemble approach, is less impacted by multicollinearity because each tree is built independently.

Here are your tasks:

  1. Build a Random Forest Regressor model using Scikit-Learn's [RandomForestRegressor](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestRegressor.html) and train it on the train data.

  2. Evaluate the performance of the model on the test data using mean absolute error as a metric. Mess around with various input parameter configurations to see how they affect the model. Can you beat the performance of the linear regression model?

**NOTE**

The above list of advantages for using Random Forest Regressor over Linear Regression in Scikit-Learn incorrectly states that the Random Forest Regressor can take in categorical and numerical datatypes. However, the regressor simply attempts to convert all datatypes into a float. If it cannot make the conversion, such as when it tries to convert a string into a float, the regressor will fail.

However, decision trees do not require data to be standardized.

Therefore, the Random Forest Regressor models below do require that the categorial feature be converted into a numerical feature (like the Linear Regression model), but it does not require that the numerical values be standardized before training the model.

**Build and Train a Random Forest Regressor Model**

Since the data set is large, the regressor model will only create 10 trees for initial testing purposes.

In [18]:
# Build random forest regressor model
rfg1 = RandomForestRegressor(n_estimators=10, random_state=42)

In [19]:
# Use get.dummies() to one-hot-encode the categorical column
new_X = pd.get_dummies(X)

In [20]:
# perform train, test, split on new_X data
X_train, X_test, y_train, y_test = train_test_split(new_X, y, test_size = 0.3, random_state=1337)

In [21]:
# fit model to training data
rfg1.fit(X_train, y_train)

RandomForestRegressor(n_estimators=10, random_state=42)

**Evaluate the model using mean absolute error**

In [22]:
# Make predictions on the test data
y_rfg_pred = rfg1.predict(X_test)

rfg1_mae = mean_absolute_error(y_test, y_rfg_pred)

In [23]:
# display results
print(f"Baseline model's mae:\t\t\t{base_mae:.5f}")
print(f"Linear Regression's mae:\t\t{linr_mae:.5f}")
print(f"First Random Forest Regressor's mae:\t{rfg1_mae:.5f}")

Baseline model's mae:			9.33219
Linear Regression's mae:		0.13891
First Random Forest Regressor's mae:	0.02225


**Adjust Hyperparameters to See Effects on Model Performance**

The following hyperparameters will be tuned individually to see how it affects the model's performance:
* min_samples_leaf
* max_features
* max_samples

These hyperparameters were chosen because they are not being tested in the upcoming section on hyperparameter tuning.

***min_samples_leaf***

In [24]:
leafs = [250, 500, 1000]

In [25]:
rfg2_scores = []

for leaf in leafs:
    rfg2 = RandomForestRegressor(n_estimators=10, random_state=42, min_samples_leaf=leaf)    
    rfg2.fit(X_train, y_train)
    y_rfg2_pred = rfg2.predict(X_test)
    rfg2_scores.append(mean_absolute_error(y_test, y_rfg2_pred))    

In [26]:
print("Min_samples_leaf:\tMAE score:")
print(f"None\t\t\t{rfg1_mae:.5f}")

i = 0

while i < len(leafs):
    print(f'{leafs[i]}\t\t\t{rfg2_scores[i]:.5f}')
    i += 1

Min_samples_leaf:	MAE score:
None			0.02225
250			0.37033
500			0.43520
1000			0.54474


***max_features***

In [27]:
feats = [11, 13, 15]

In [28]:
rfg3_scores = []

for feature in feats:
    rfg3 = RandomForestRegressor(n_estimators=10, random_state=42, max_features=feature)    
    rfg3.fit(X_train, y_train)
    y_rfg3_pred = rfg3.predict(X_test)
    rfg3_scores.append(mean_absolute_error(y_test, y_rfg3_pred))  

In [29]:
print("Max_features:\tMAE score:")
print(f"None\t\t\t{rfg1_mae}")

i = 0

while i < len(feats):
    print(f'{feats[i]}\t\t\t{rfg3_scores[i]}')
    i += 1

Max_features:	MAE score:
None			0.022253336566018755
11			0.03941806935497831
13			0.03198935206291044
15			0.027244393078914037


***max_samples***

In [30]:
max_samps = [0.6, 0.7, 0.8]

In [31]:
rfg4_scores = []

for samples in max_samps:
    rfg4 = RandomForestRegressor(n_estimators=10, random_state=42, max_samples=samples)    
    rfg4.fit(X_train, y_train)
    y_rfg4_pred = rfg4.predict(X_test)
    rfg4_scores.append(mean_absolute_error(y_test, y_rfg4_pred))  

In [32]:
print("Max_samples:\tMAE score:")
print(f"None\t\t\t{rfg1_mae:.5f}")

i = 0

while i < len(max_samps):
    print(f'{max_samps[i]}\t\t\t{rfg4_scores[i]:.5f}')
    i += 1

Max_samples:	MAE score:
None			0.02225
0.6			0.02540
0.7			0.02456
0.8			0.02333


**Performance of all tested models**

In [33]:
# display results
print(f"MAE SCORES\nBaseline model:\t\t\t\t{base_mae:.5f}")
print(f"Linear Regression:\t\t\t{linr_mae:.5f}")
print(f"\nBest Random Forest Regressor:\n  set to 10 n_estimators\t\t{rfg1_mae:.5f}")

MAE SCORES
Baseline model:				9.33219
Linear Regression:			0.13891

Best Random Forest Regressor:
  set to 10 n_estimators		0.02225


When comparing mae scores, both the Linear Regression and Random Forest models performed significantly better than the baseline model. 

Between the Linear Regression and Random Forest models, the Random Forest models all had a lower mae score.

None of the hyperparameters tuned above helped improve the mae scores of the Random Forest Regressor models. The best mae score was always 0.02225, which comes from the baseline Random Forest Regressor model. This model only set the hyperparameter of `n_estimators` to 10.

### Hyperparameter Tuning

Hyperparameter tuning plays a critical role in machine learning model development. It involves selecting the optimal values for the hyperparameters, which are configuration settings that control the behavior of the learning algorithm. Here's why hyperparameter tuning is so important in ML:

1. **Optimizing Model Performance**: The choice of hyperparameters can significantly impact the model's performance. By fine-tuning the hyperparameters, we can improve the model's accuracy, precision, recall, or other performance metrics. It helps to extract the maximum predictive power from the chosen algorithm and ensures that the model is well-suited to the specific problem at hand.

2. **Avoiding Overfitting and Underfitting**: Hyperparameter tuning helps strike a balance between overfitting and underfitting.

3. **Exploring Model Complexity**: Hyperparameter tuning enables us to explore the complexity of the model. For instance, in algorithms like decision trees or neural networks, we can adjust the number of layers, the number of neurons, or the maximum depth of the tree. By systematically modifying these hyperparameters, we can understand how different levels of complexity impact the model's performance and find the right balance between simplicity and complexity.

Note, there are multiple approaches to hyperparemeter tuning.  

While grid search is the easiest to understand and implement there are many advantages of Bayesian search over grid search for hyperparameter tuning:

1. **Efficiency**: Bayesian search is generally more efficient than grid search. Grid search explores all possible combinations of hyperparameter values, which can be computationally expensive and time-consuming, especially when dealing with a large number of hyperparameters or a wide range of values. Bayesian search, on the other hand, intelligently selects the next hyperparameter configuration to evaluate based on the results of previous evaluations. It focuses on areas of the hyperparameter space that are more likely to yield better performance, reducing the number of evaluations needed.

2. **Flexibility**: Bayesian search is flexible in handling continuous and discrete hyperparameters. It can handle both types of hyperparameters naturally and effectively. In contrast, grid search is more suitable for discrete hyperparameters but may struggle with continuous ones, as it requires discretization or defining a finite set of values to search over.

3. **Adaptive Search**: Bayesian search adapts its search strategy based on the results of previous evaluations. It maintains a probability distribution over the hyperparameter space, updating it with each evaluation. This allows it to dynamically allocate more evaluations to promising regions and explore unexplored areas. In contrast, grid search follows a fixed and predefined search grid, regardless of the results of previous evaluations.

4. **Better Convergence**: Bayesian search has the potential to converge to the optimal hyperparameter configuration more quickly.

Here are your tasks:

  1. Perform a grid-search on a Random Forest Regressor model. Only search the space for the parameters 'n_estimators', 'max_depth', and 'min_samples_split'. Note, this can take some time to run. Make sure you set reasonable boundaries for the search space. Use Scikit-Learn's [GridSearchCV](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html) method.

  2. After you've identified the best parameters, train a random forest regression model using these parameters on the full training data.

  3. Evaluate the model from the previous step using the test data. How does your model perform?

**Define Hyperparameters**

The following hyperparameters will be tested and the values used during testing explained.

* **n_estimators** - random forests tend to perform better when more trees are used to make the predictions, helping reduce overfitting and underfitting
  * values to test: 50, 150, 250, 350
  
* **max_depth** - a max_depth between 4 and 8 is usually the ideal depth for the trees created, as it balances the size of the tree so they fall between overly simplified (underfitting) and overly complex (overfitting) decision trees
  * values to test: None, 5, 6, 7, 8
  
* **min_samples_split** - having a low number for min_samples_split may cause a random forest tree to overfit the data as a node can have as little as two samples in it to split the node into two. Using a value of 1% or 2% of the training sample size may help reduce overfitting and speed up the model training time
  * values to test: 2 (the default split value), 1% and 2% of training size

In [34]:
one_percent = int(round(len(X_train) * 0.01, 0))
two_percent = int(round(len(X_train) * 0.02, 0))

# Define the hyperparameters to tune.
param_grid = {
    'n_estimators': [50, 150, 250, 350],
    'max_depth': [None, 5, 6, 7, 8],
    'min_samples_split': [2, one_percent, two_percent]
}

In [35]:
# Perform grid search to find the best hyperparameters. This could take a while.
grid_search = GridSearchCV(RandomForestRegressor(random_state=42), param_grid, cv=3, n_jobs=-1)
grid_search.fit(X_train, y_train)

GridSearchCV(cv=3, estimator=RandomForestRegressor(random_state=42), n_jobs=-1,
             param_grid={'max_depth': [None, 5, 6, 7, 8],
                         'min_samples_split': [2, 16747, 33494],
                         'n_estimators': [50, 150, 250, 350]})

In [36]:
# Get the best model and its parameters.
best_model = grid_search.best_estimator_

print(f'Best regressor:\n{best_model}')
print(f'\nHyperparameters set on best regressor:\n{grid_search.best_params_}')
print(f'\nAll hyperparameters of best regressor:\n{best_model.get_params()}')

Best regressor:
RandomForestRegressor(n_estimators=350, random_state=42)

Hyperparameters set on best regressor:
{'max_depth': None, 'min_samples_split': 2, 'n_estimators': 350}

All hyperparameters of best regressor:
{'bootstrap': True, 'ccp_alpha': 0.0, 'criterion': 'squared_error', 'max_depth': None, 'max_features': 1.0, 'max_leaf_nodes': None, 'max_samples': None, 'min_impurity_decrease': 0.0, 'min_samples_leaf': 1, 'min_samples_split': 2, 'min_weight_fraction_leaf': 0.0, 'n_estimators': 350, 'n_jobs': None, 'oob_score': False, 'random_state': 42, 'verbose': 0, 'warm_start': False}


In [37]:
# import time to calculate the time it takes to train, fit, and make predictions with the model
import time

In [38]:
# train model and track time to execute
start_time = time.time()

best_model.fit(X_train, y_train)

end_time = time.time()
execution_time = end_time - start_time

In [39]:
# display time it took to fit model on all training data
print("Execution time:",execution_time)

Execution time: 2270.7194213867188


In [40]:
# Make predictions on the test data
start_time = time.time()

y_rfgb_pred = best_model.predict(X_test)

end_time = time.time()
execution_time = end_time - start_time

In [41]:
# display time it took to make predictions with the model
print("Execution time:",execution_time)

Execution time: 40.18590521812439


In [42]:
# calculate mean absolute error on the predictions
best_rfg_mae = mean_absolute_error(y_test, y_rfgb_pred)

## Hyperparamter Results

In [54]:
# print results of all models
# display results
print(f"Random Forest Regressor MAE Scores")
print(f"  10 estimators:\t\t{rfg1_mae:.5f}")
print(f"  350 estimators:\t\t{best_rfg_mae:.5f}")

Random Forest Regressor MAE Scores
  10 estimators:		0.02225
  350 estimators:		0.02028


The only parameter requiring tuning to improve the mae score of the Random Forest Regressor model was n_estimators. The best score came from the highest n_estimators tested: 350. However, the best model's mae score was only slightly better than the baseline Random Forest Regressor's score -- the difference was less than 0.002. 

It took the best model about 2,270.7 seconds (or 37 minutes and 50.7 seconds) to train and fit against the test data. It also took the model about 40.2 seconds to make predictions for all the X_test samples. 

Let's look at how long it took to fit and predict with the baseline Random Forest Regressor model.

In [46]:
# train model and display time to execute
rfg1 = RandomForestRegressor(n_estimators=10, random_state=42)

In [47]:
start_time = time.time()

rfg1 = RandomForestRegressor(n_estimators=10, random_state=42)
rfg1.fit(X_train, y_train)

end_time = time.time()
execution_time = end_time - start_time

In [48]:
print("Execution time:",execution_time)

Execution time: 64.49794435501099


In [50]:
# Make predictions on the test data
start_time = time.time()

y_rfg_pred = rfg1.predict(X_test)

end_time = time.time()
execution_time = end_time - start_time

In [51]:
print("Execution time:",execution_time)

Execution time: 1.1883256435394287


For the basline Random Forest Regressor model, it took just over a minute to train and fit the model on the training data. It took just over a second to then predict the y values for all the X_test samples.

# Summary

In [52]:
# print results of all models
# display results
print(f"MAE SCORES\nBaseline model:\t\t\t\t{base_mae:.5f}")
print(f"Linear Regression:\t\t\t{linr_mae:.5f}")
print(f"\nRandom Forest Regressors with:\n  10 estimators:\t\t\t{rfg1_mae:.5f}")
print(f"  best model, 350 estimators:\t\t{best_rfg_mae:.5f}")

MAE SCORES
Baseline model:				9.33219
Linear Regression:			0.13891

Random Forest Regressors with:
  10 estimators:			0.02225
  best model, 350 estimators:		0.02028


Both the Linear Regression and Random Forest Regressor models perform significantly better than the baseline model when comparing each model's mean absolute error (mae) score.

The Random Forest Regressor performs better than the Linear Regression model, as the mae score for the Linear Regression model was 0.13891 while the best Random Forest Regressor model had a mae of 0.02028. A difference of almost 0.12.

However, the difference between the best Random Forest Regressor's mae score and the baseline random forest model was very small. The difference in scores was less than 0.002.

Since the baseline random forest regressor model only needs 10 estimators, it took just over a minute to fit the model to the training data and just over a second to make predictions for all of the X_test samples. On the other hand, since the best random forest regressor model uses 350 estimators, it took almost 40 times as long to fit and make predictions.

If attempting to find the best model, it would be helpful to look at other metrics for these differing Random Forest Regressor models to determine if the increase in time and resources to create the best model is worth the increase in performance as compared to using a model with much fewer estimators.